### DISCUSSION TITLE
Trying to upgrade to SQLAlchemy 2.0.0b4 and having issues with `load_only` `attrs` typing

### DISCUSSION THREAD

**Discussion by shiraSC** [sqlalchemy/sqlalchemy#9010] [Usage Questions] [answered]
Hello,

I'm attempting to upgrade to SQLAlchemy 2.0.0b4 and having issues with `load_only` `attrs` typing. In the documentation it says that `load_only` accepts either `str` or `QueryableAttribute[Any]`. https://docs.sqlalchemy.org/en/20/orm/queryguide/columns.html#sqlalchemy.orm.load_only

Our code calls `load_only` with strings and I am getting this error:

```
web-dynamic_1           | Traceback (most recent call last):
web-dynamic_1           |   File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/strategy_options.py", line 2345, in _parse_attr_argument
web-dynamic_1           |     insp: InspectionAttr = inspect(attr)  # type: ignore
web-dynamic_1           |   File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/inspection.py", line 118, in inspect
web-dynamic_1           |     raise exc.NoInspectionAvailable(
web-dynamic_1           | sqlalchemy.exc.NoInspectionAvailable: No inspection system is available for object of type <class 'str'>
web-dynamic_1           | 
web-dynamic_1           | The above exception was the direct cause of the following exception:

........

web-dynamic_1           |  query = query.options(load_only('string1', 'string2', string3', 'string4'))
web-dynamic_1           |   File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/strategy_options.py", line 2398, in load_only
web-dynamic_1           |     _, lead_element, _ = _parse_attr_argument(attrs[0])
web-dynamic_1           |   File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/strategy_options.py", line 2347, in _parse_attr_argument
web-dynamic_1           |     raise sa_exc.ArgumentError(
web-dynamic_1           | sqlalchemy.exc.ArgumentError: expected ORM mapped attribute for loader strategy argument
```
The variable `query` is set further up to: `query = db_session.query(cls.model)`

`load_only` was working in 1.4.45 and this only broke now in 2.0.0b4. The documentation for `load_only` doesn't seem to have changed (aside from adding the typing) but I'm wondering if I need to do something different to use it with string arguments?

Thanks in advance,
Shira

**Comment by zzzeek** [comment, marked as answer]
hi and thanks for trying SQLAlchemy 2.0!

lets try a program that has the use which you describe:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import load_only
from sqlalchemy.orm import Session
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    x = Column(String)
    y = Column(String)

e = create_engine("sqlite://")
Base.metadata.create_all(e)

s = Session(e)

s.query(A).options(load_only("x")).all()

```

this program succeeds on 1.4 .  When we want to upgrade to 2.0, follow the process at [Migrating to SQLAlchemy 2.0](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#migrating-to-sqlalchemy-2-0).   Once we have SQLAlchemy 1.4 working, the next step is [Turn on RemovedIn20Warnings](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#migration-to-2-0-step-two-turn-on-removedin20warnings).   Running the above program with this enabled looks like:

```
$ SQLALCHEMY_WARN_20=1 python test3.py 
/home/classic/dev/sqlalchemy/test3.py:23: RemovedIn20Warning: Using strings to indicate column or relationship paths in loader options is deprecated and will be removed in SQLAlchemy 2.0.  Please use the class-bound attribute directly. (Background on SQLAlchemy 2.0 at: https://sqlalche.me/e/b8d9)
  s.query(A).options(load_only("x")).all()
```


so we can see above the line of code that needs a change is noted by the warning.  This specific change is noted at https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#orm-query-joining-loading-on-relationships-uses-attributes-not-strings , which currently refers to the joinedload() option, but applies to other load options as well.

Additionally, the documentation *has* been mostly rewritten for options like these; make sure you are looking at the 2.0 documentation at https://docs.sqlalchemy.org/en/20/.   At this point SQLAlchemy 2.0's documentation is *mostly* up to date for the newer APIs, so you can also take a look newly rewritten guide for deferred column loading at https://docs.sqlalchemy.org/en/20/orm/queryguide/columns.html#using-load-only-to-reduce-loaded-columns .  it should be linked from the load_only() docstring.

**Reply by shiraSC** [reply]
Wow! Thank you for your speedy reply! I'll see what I can do to fix this on my end.
